Pipeline/PackageBuildOutputs/utilities/sbomGenerator.groovy

```diff
@@ -0,0 +1,206 @@
+@groovy.transform.BaseScript com.ibm.dbb.groovy.ScriptLoader baseScript
+import java.io.File
+import com.ibm.dbb.build.CopyToHFS
+import com.ibm.dbb.build.BuildProperties
+import com.ibm.dbb.build.DBBConstants
+import com.ibm.dbb.build.DBBConstants.CopyMode
+import com.ibm.dbb.build.report.BuildReport
+import com.ibm.dbb.build.report.records.*
+import groovy.transform.*
+import groovy.cli.commons.*
+import java.nio.file.*
+import static java.nio.file.StandardCopyOption.*
+import com.ibm.jzos.ZFile
+import org.cyclonedx.*
+import org.cyclonedx.model.*
+
+/************************************************************************************
+ * This utility script helps generating an SBM from the DBB BuildReport
+ *
+ * Version 1 - 04/2024
+ *  Initial implementation of SBOM Generation
+ ************************************************************************************/
+
+
+// Property prefix, matching what is defined in zAppBuild/build.groovy
+@Field String hashPrefix = ':githash:'
+@Field bomRef = 0
+
+// SBOM generation-related objects
+@Field HashSet<String> applications
+@Field HashMap<String, HashMap<String, String>> outputsInSBOM
+@Field ArrayList<Dependency> sbomDependencies
+@Field Bom sbom
+
+def initializeSBOM(String sbomAuthor) {
+	// HashMap of files to report in SBOM
+	applications = new HashSet<String>()
+	outputsInSBOM = new HashMap<String, HashMap<String, String>>()
+	sbom = new Bom();
+	sbom.setSerialNumber("url:uuid:" + UUID.randomUUID().toString());
+	sbom.setVersion(1);
+	LifecycleChoice sbomLifecycleChoice = new LifecycleChoice()
+	sbomLifecycleChoice.setPhase(LifecycleChoice.Phase.POST_BUILD)
+	Lifecycles sbomLifecycles = new Lifecycles()
+	sbomLifecycles.setLifecycleChoice([sbomLifecycleChoice])
+	Metadata sbomMetadata = new Metadata()
+	sbomMetadata.setLifecycles(sbomLifecycles)
+	// Add the Author is passed on command-line, and well-formed ("Name <email>" expected)
+	if (sbomAuthor) {
+		def sbomAuthorFields = sbomAuthor.split('<')
+		if (sbomAuthorFields.size() == 2) {
+			OrganizationalContact author = new OrganizationalContact()
+			author.setName(sbomAuthorFields[0].trim())
+			author.setEmail(sbomAuthorFields[1].replaceAll(">", "").trim())
+			sbomMetadata.addAuthor(author)
+		} else {
+			println("*! Warning: SBOM Author not correctly formed, expecting 'Name <email>' format. Skipping.")
+		}
+	} else {
+		println("*! Warning: empty SBOM Author. It is recommend to specify a valid Author.")	
+	}  
+	sbom.setMetadata(sbomMetadata)
+    sbom.setDependencies(new ArrayList<Dependency>())
+}
+
+def addEntryToSBOM(DeployableArtifact deployableArtifact, HashMap<String, String> info) {
+
+	String container = info.get("container")
+	String owningApplication = info.get("owningApplication")
+	Record record = info.get("record")
+	PropertiesRecord propertiesRecord = info.get("propertiesRecord")
+	DependencySetRecord dependencySetRecord = info.get("dependencySetRecord")
+			
```

Comment:
Can you please add some comments what the sections are doing?

Reply:
Done